File: algorithms/sorting/quick-sort/QuickSort.java
public class QuickSort {
  public static int partition(int[] arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++) {
      if (arr[j] < pivot) {
        i++;
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
      }
    }

    i++;
    int temp = arr[i];
    arr[i] = pivot;
    arr[high] = temp;

    return i;
  }

  public static void quickSort(int arr[], int low, int high) {
    if (low < high) {
      int pivot_index = partition(arr, low, high);

      quickSort(arr, low, pivot_index - 1);
      quickSort(arr, pivot_index + 1, high);
    }
  }
}

File: algorithms/sorting/quick-sort/QuickSort.ts
class Sorting {
	private partition(array: number[], low: number, high: number): number {
		const pivot: number = array[high];
		let i: number = low - 1;
		let temp: number;

		for (let j: number = low; j < high; j++) {
			if (array[j] < pivot) {
				i++;
				temp = array[i];
				array[i] = array[j];
				array[j] = temp;
			}
		}

		temp = array[i + 1];
		array[i + 1] = array[high];
		array[high] = temp;
		return i + 1;
	}

	public QuickSort(
		unsortedArray: number[],
		low: number,
		high: number,
	): number[] {
		if (low < high) {
			const pivotIndex: number = this.partition(unsortedArray, low, high);
			this.QuickSort(unsortedArray, low, pivotIndex - 1);
			this.QuickSort(unsortedArray, pivotIndex + 1, high);
		}

		return unsortedArray;
	}
}

export { Sorting };
